[RFC] glfdc expression evaluator for runtime shape inference

prak-nv opened the pull request:
This PR adds runtime expression evaluator. The expressions are memoized during compilation phase to be later evaluated during rutime phase, when all exact values are known. This change is still work in progress, the area I'll be focussing on mostly is is evaluator itself (content of glfdc directory). I will be benchmarking and possibly tunning it and as well improving code readability and tighter integration with pytorch.

prak-nv committed:
Allow testing if const kir::Val* is input/output

This change makes possible to use those member functions from
inmutable passes and IR visitors.

prak-nv committed:
Glue in shape infer expression eval prototype.

prak-nv commented on torch/csrc/jit/codegen/cuda/executor_utils.cpp:
```diff
@@ -523,7 +525,10 @@ kir::ExpressionEvaluator bindKernelInputs(
+  auto mapping = glfdc::ReusedExprMapping::create_lazy_mapping(
+      kernel->summary().expr_reuses);
+  std::unique_ptr<glfdc::EvalState> es{new glfdc::EvalState{mapping}};
+  kir::ExpressionEvaluator expr_eval{std::move(es)};
```
Needs better API

prak-nv replied:
Expression reuses bitmap was moved from ExprBuilder into ExprDAG; thus we don't need to deal with it and creating EvalState is simpler. I've added helper createEvalStateForDAG() that handles it right now.

naoyam commented:
I'm going to review this PR. For the meantime, please consider adding tests. It might make sense to have a new test file, like "test_gpu_glfdc.cpp", in the test directory.

naoyam commented on torch/csrc/jit/codegen/cuda/expr_evaluator.cpp:
```diff
@@ -5,6 +5,8 @@
 #include <torch/csrc/jit/codegen/cuda/ir_all_nodes.h>
 #include <torch/csrc/jit/codegen/cuda/ir_iostream.h>
 
+#include <torch/csrc/jit/codegen/cuda/glfdc/expr_builder.h>
```
This doesn't seem necessary.

prak-nv replied:
Removed in updated version.

naoyam commented on torch/csrc/jit/codegen/cuda/glfdc/eval.h:
```diff
@@ -0,0 +1,169 @@
+
+#include "c10/util/Optional.h"
+
+namespace glfdc {
```
Please place this name space inside our umbrella namespace, `torch::jit::fuser::cuda`.

prak-nv replied:
Will do

prak-nv replied:
All of glfdc is moved to `torch::jit::fuser::cuda::glfdc` updated version

naoyam commented on torch/csrc/jit/codegen/cuda/glfdc/eval.h:
```diff
@@ -0,0 +1,169 @@
+
+// Expression evaluator
+
+#include "bitvector.h"
```
This should be `<torch/csrc/jit/codegen/cuda/glfdc/bitvector.h>` to follow the convention in PyTorch.

prak-nv replied:
Will do

prak-nv replied:
Done and all other should be ok right now.

naoyam commented on torch/csrc/jit/codegen/cuda/shape_expr_memo.h:
```diff
@@ -0,0 +1,152 @@
+      current_operand_.has_value() && "Unexpected operator type");
+}
+
+inline void ExprBuilderVisitor::visit(const BinaryOp* binary_op) {
```
I'd prefer functions like this placed in a source file. The possible gain by inlining would be negligible compared to the time to execute the function body.

prak-nv replied:
Sure

prak-nv replied:
All moved to cpp file

naoyam replied:
I still see some heavy header files in the glfdc directory.

naoyam commented:
Some general comments:

- Our naming convention is documented here: https://docs.google.com/document/d/14c2wNOmGoolpVjsiYcEBTbALkEzrqaTppGoJ5NkiAV0/edit. It's not strictly enforced, especially in some old code, but please follow the guideline as much as possible.
- I see most of the functions are defined in header files. Most of them can be placed in source files with negligible performance costs.
- Run clang-format. For instructions, see https://docs.google.com/document/d/1F2-ZtadMz1NliEZmFRvtOvfosRN5N337sGsAGUu3VSM/edit#heading=h.hpz1l54zbzbh.

naoyam commented on torch/csrc/jit/codegen/cuda/executor.cpp:
```diff
@@ -274,6 +274,47 @@ at::Tensor inferAndAlloc(
   }
 }
 
+at::Tensor inferAndAlloc(
```
Is the old `inferAndAlloc` going to be removed? I see it's still used for allocating outputs.

prak-nv replied:
Old inferAndAlloc is still used for static allocations, where it didn't make sense to memoize expression.

naoyam replied:
The two versions look quite similar. Can they be merged?

prak-nv replied:
Merged those functions. Please let me know if it looks ok in your eyes.

naoyam commented on torch/csrc/jit/codegen/cuda/executor.cpp:
```diff
@@ -329,44 +370,65 @@ uint64_t FusionExecutor::computeSharedMemory(
+  for (auto smem_alloc : buffers) {
+    if (smem_alloc->alias() == nullptr) {
+      auto mem_inferred = expr_eval.evaluateMemoized(memoized[i]);
+      const auto inferred_val = expr_eval.evaluate(smem_alloc->size());
```
Do we still need to look at `smem_alloc->size()`? Can we just use `memoized[i]`?

prak-nv replied:
No, its a leftover, I did not remove sorry.

prak-nv replied:
Should be fixed

naoyam commented on torch/csrc/jit/codegen/cuda/executor.cpp:
```diff
@@ -303,7 +344,7 @@ uint64_t FusionExecutor::computeSharedMemory(
     bool align_padding,
     uint64_t total) {
-  FUSER_PERF_SCOPE("computeSharedMemory");
+  FUSER_PERF_SCOPE("computeSharedMemory::old");
```
Do we need to keep this old version?

prak-nv replied:
Will try to do that in follow up change, if performance regression wouldn't be too bad. Other thing would be getting rid of all uses of old evaluator (building expression in place and evaluating them in place).

prak-nv replied:
This is still work in progress, not included in current version

naoyam commented on torch/csrc/jit/codegen/cuda/executor.cpp:
```diff
@@ -456,11 +522,21 @@ FusionExecutor::GlobalBuffers FusionExecutor::allocGlobalVals(
+      global_buffers.empty_buffers.push_back(inferAndAlloc(
+          tv,
+          alloc->shape(),
+          kernel_summary.memoized_global_shapes[i - 1],
```
nit: `i-1` -> `i` and move `++i` to the end of the loop body.

prak-nv replied:
Done.

naoyam commented on torch/csrc/jit/codegen/cuda/kernel.h:
```diff
@@ -15,19 +17,32 @@ namespace fuser {
+
 //! Summary of interesting facts about the kernel
 struct KernelSummary {
+  std::unique_ptr<glfdc::ExprDAG> memoized_dag;
```
Please comment the new fields.

naoyam replied:
Where is `memoized_dag` used?

prak-nv replied:
Every memoized expression (glfdc::Expr and later glfdc::ExprEvaluator) refers to the part of DAG (but not owns) , thus we have to ensure it is alive.
Reference counting schemes in that case would be overkill IMO.

prak-nv replied:
Added comments

naoyam replied:
Thanks. Pease add a comment to `memoized_dag` as well.

naoyam commented on torch/csrc/jit/codegen/cuda/kernel.h:
```diff
@@ -15,19 +17,32 @@ namespace fuser {
   //! List of global buffers
   std::vector<const kir::Allocate*> global_allocations;
+  std::vector<glfdc::ExprEvaluator> memoized_global_allocations;
+  std::vector<std::vector<glfdc::ExprEvaluator>> memoized_global_shapes;
```
Do we need to have both?

prak-nv replied:
No, we do not

prak-nv replied:
Removed

naoyam commented on torch/csrc/jit/codegen/cuda/kernel.cpp:
```diff
@@ -231,23 +233,82 @@ class ValidateAllocation : private kir::IrVisitor {
+
+    for (auto shape_expr : allocate->shape()) {
+      auto expr = builder_visitor_->build(shape_expr);
+      TORCH_CHECK(expr.has_value());
```
Is it guaranteed to has a value? If not, please add an error message to the assertion.

prak-nv replied:
Return type of build method doesn't need to be optional. Moreover it can return expression in memoized form to avoid additional steps.

prak-nv replied:
This shouldn't be concern in current version as build return result is not any longer c10::optional<Expr>.

naoyam commented on torch/csrc/jit/codegen/cuda/kernel.cpp:
```diff
@@ -231,23 +233,82 @@ class ValidateAllocation : private kir::IrVisitor {
+      case MemoryType::Shared:
+        if (!ExpressionEvaluator::isConst(allocate->size())) {
+          auto expr = builder_visitor_->build(allocate->size());
+          TORCH_CHECK(expr.has_value());
```
Same here. Guaranteed to has a value?

prak-nv replied:
As above

naoyam commented on torch/csrc/jit/codegen/cuda/kernel.cpp:
```diff
@@ -231,23 +233,82 @@ class ValidateAllocation : private kir::IrVisitor {
+      case MemoryType::Global: {
+        auto expr = builder_visitor_->build(allocate->size());
+        if (expr.has_value())
+          summary_->memoized_global_allocations.emplace_back(expr.value());
```
Do we need this? `memoized_global_allocations` doesn't seem to be used.

prak-nv replied:
It's not used, will remove it

prak-nv replied:
Removed in current version

naoyam commented on torch/csrc/jit/codegen/cuda/lower2device.cpp:
```diff
@@ -349,12 +351,62 @@ void GpuLower::lower() {
+    if (val->getValType().value() == ValType::TensorView) {
+      used_tvs.push_back(val->as<TensorView>());
+    }
+  }
```
```suggestion
 auto used_vals = fusion_->usedMathVals();
 auto used_tvs = ir_utils::filterByType::<TensorView>(used_vals);
```

prak-nv replied:
Thanks

prak-nv replied:
Also  #998 for setUsedTVs

prak-nv replied:
Done in current version

naoyam commented on torch/csrc/jit/codegen/cuda/lower2device.cpp:
```diff
@@ -349,12 +351,62 @@ void GpuLower::lower() {
+        }
+      }
+    }
+  }
```
This seems to make sense to me, but I wonder why it was previously done at the launch time. Even without the new expression evaluator, we could just cache `kir::Val*` for each parallel type so that at the launch time we wouldn't need to analyze the kernel every time. 

@jjsjann123, do you have any concern here? Specifically, this part of the code is now moved at the lowering time:

https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/executor.cpp#L339-L355

jjsjann123 replied:
Looks like the code is just moved to a different place but the logic follows the same.

IIRC, `computeLaunchParams` is used in kernel segmentation as well, but I don't think their parallel type binding would change so @naoyam's suggestion here for caching that sounds valid. cc'ing @shmsong

shmsong replied:
Makes sense. Caching the `Val`'s here will save one usedTV traversal on the runtime. Segmenter doesn't look at the parallel types and launch parameters, or at least not yet. But I can add caching of the parallelized extents right after each segment is scheduled if that sounds like a good idea. 

Also another thing to note is that launch parameters are re-computed only if we have an input miss but we have a heuristic hit, which is the second level hit in our caching system. The fastest code path where we have exact input dimension match wouldn't need to compute launch params. 

On a separate future thread when we revisit run time. We probably need more profiling effort to see how important is the speed of the heuristic hit, and how important `computeLaunchParams` is in heuristic hit delay.

naoyam replied:
Thanks @jjsjann123 and @shmsong. I feel more confident with this change.

naoyam commented on torch/csrc/jit/codegen/cuda/lower2device.cpp:
```diff
@@ -430,19 +481,33 @@ class GpuLower::KernelIrMapper : private OptInConstDispatch {
+
+          std::string("arg") + KirT::scalarTypePrefix() +
+              std::to_string(node->name()),
+          node->getDataType().value());
```
Have we decided to do this? Was there some concerns? Pinging @shmsong

prak-nv replied:
If we're not I can also prepare version without it.

prak-nv replied:
Should I prepare version without this change comparison?

shmsong replied:
I think it should be ok if only scalar input is named and the scalar's name is automatically generated. We'll probably need to keep track of the created names if we keeping creating different kinds of `NamedScalar`'s later. 

@prak-nv could you show an example of what the generated kernel (maybe just a snippet) looks like with this change?

piotrrak replied:
Will paste it morning.

prak-nv replied:
I am struggling right now to get my exact case (int argument dependend kir::Allocation) after rebase, but it is basically how it looks like:

```
[ RUN      ] NVFuserTest.FusionScalarInputs_CUDA

KERNEL (T0, T1, argd8, argd9, argd10, argd11) -> (T4) :
  DEFINE_MAGIC_ZERO
  d12 = ALLOCATE(mem_type=register, size=1, zero_init=false)
  d12 = argd8 * argd9
  d13 = ALLOCATE(mem_type=register, size=1, zero_init=false)
  d13 = argd10 - argd11
  ~ i24 = T0.size[0] * T0.size[1]
  ~ i26 = ceilDiv(i24, 128)
  ~ i29 = ceilDiv(i26, 4)
  FOR ki96 in id18[blockIdx.x.i(0 .. i29)]:
...
```
Will try to catch it later with sophisticated enough asserts.

prak-nv replied:
After:
```
[ RUN      ] NVFuserTest.FusionSmemDynamicPersistentSoftmax2D_CUDA

KERNEL (T0, argi10) -> (T7) :
  ~ i14 = ceilDiv(T0.size[1], argi10)
  ~ ki142 = i14 * argi10
  T8 = ALLOCATE(mem_type=shared, size=ki142, zero_init=false)
  ~ i23 = ceilDiv(T0.size[1], argi10)
  ~ ki146 = i23 * argi10
  T4 = ALLOCATE(mem_type=shared, size=ki146, zero_init=false)
      .alias=T8
```
Before:
```
[ RUN      ] NVFuserTest.FusionSmemDynamicPersistentSoftmax2D_CUDA

KERNEL (T0, i10) -> (T7) :
  ~ i14 = ceilDiv(T0.size[1], i10)
  ~ ki142 = i14 * i10
  T8 = ALLOCATE(mem_type=shared, size=ki142, zero_init=false)
  ~ i23 = ceilDiv(T0.size[1], i10)
  ~ ki146 = i23 * i10
  T4 = ALLOCATE(mem_type=shared, size=ki146, zero_init=false)
      .alias=T8
```

naoyam commented on torch/csrc/jit/codegen/cuda/shape_expr_memo.h:
```diff
@@ -0,0 +1,152 @@
+
+  glfdc::Operand build_operand(const Val* value) {
+    TORCH_CHECK(value->isScalar());
+    TORCH_CHECK(value->dtype() == DataType::Int);
```
nit: Use `TORCH_INTERNAL_ASSERT` if these assertions are to validate the internal logic. `TORCH_CHECK` is for validating user inputs. It should be always used with an error message that's understandable by the user.

prak-nv replied:
Sure, think I changed most of them, will be still tweeking messages for TORCH_INTERNAL_ASSERT pending review.

naoyam commented on torch/csrc/jit/codegen/cuda/shape_expr_memo.h:
```diff
@@ -0,0 +1,152 @@
+  if (auto def = value->definition()) {
+    def->accept(this);
+  } else if (kernel_->isInput(value)) {
+    TORCH_CHECK(false && "Unhandled input scalar");
```
```suggestion
    TORCH_INTERNAL_ASSERT(false, "Unhandled input scalar");
```

naoyam commented on torch/csrc/jit/codegen/cuda/shape_expr_memo.h:
```diff
@@ -0,0 +1,152 @@
+namespace cuda {
+namespace kir {
+
+class ExprBuilderVisitor : IrVisitor {
```
Please comment the class.

prak-nv replied:
Done

naoyam commented on torch/csrc/jit/codegen/cuda/kernel_expr_evaluator.cpp:
```diff
@@ -3,14 +3,166 @@
+  void visit(const UnaryOp* unary_op) final;
+  void visit(const BinaryOp* binary_op) final;
+
+  glfdc::ExpressionBuilder* builder_;
```
nit: `builder_ = nullptr`, or make it a reference.

prak-nv replied:
This code is removed

naoyam commented on torch/csrc/jit/codegen/cuda/kernel_expr_evaluator.h:
```diff
@@ -37,13 +43,19 @@ class TORCH_CUDA_CU_API ExpressionEvaluator : private IrVisitor {
 
   //! Try to evaluate a Kernel IR value
   c10::optional<Int::ScalarType> evaluate(const Val* value);
+  c10::optional<Int::ScalarType> evaluateMemoized(const glfdc::ExprEvaluator&);
```
Why don't we simply call this `evaluate`?

prak-nv replied:
Renamed

naoyam commented on torch/csrc/jit/codegen/cuda/kernel_expr_evaluator.cpp:
```diff
@@ -34,14 +197,12 @@ c10::optional<Int::ScalarType> ExpressionEvaluator::evaluate(const Val* value) {
-
   const auto post_eval_it = known_values_.find(value);
   return post_eval_it != known_values_.end()
       ? c10::optional<Int::ScalarType>(post_eval_it->second)
```
Why did you modify this function? As far as I understand, the use of the new expression evaluator shouldn't affect this interface, right?

prak-nv replied:
Just me being sloppy removing some additional asserts I had for development purposes.

naoyam commented on torch/csrc/jit/codegen/cuda/glfdc/sexpr.h:
```diff
@@ -0,0 +1,128 @@
+
+// Symbolic operand of subexpression
+struct UnboundValue {
+  std::size_t index_;
```
What does this `index_` mean?

prak-nv replied:
Added comment, please let me know is it clear enough now.

naoyam commented on torch/csrc/jit/codegen/cuda/glfdc/sexpr.h:
```diff
@@ -0,0 +1,128 @@
+
+// Reference to subexpression (index to subexpression DAG)
+struct SExprRef {
+  std::size_t index_;
```
same here

prak-nv replied:
Ditto

naoyam commented on torch/csrc/jit/codegen/cuda/glfdc/expr_builder.h:
```diff
@@ -0,0 +1,79 @@
+  sexpr_lookup_t seen_exprs_;
+
+  // only needed for lazy_eval construction
+  bitvector_t reused_subexpressions_;
```
What does this "reuse" mean? What is it for?

prak-nv replied:
During evaluation EvalState will remember values of reused subexpression.

Improved comment, its bitmap meant for CSE during evaluation. It makes no sense to remember calculated values of subexpressions if those are not reused.

naoyam commented:
I'm getting confused with the expression-related classes in `namespace glfdc`. Can you explain what they represent and how they should be used?

- glfdc::Expr
- glfdc::ExprDag
- glfdc::SExpr
- glfdc::SExprRef

prak-nv commented on torch/csrc/jit/codegen/cuda/kernel_expr_evaluator.cpp:
```diff
@@ -3,14 +3,166 @@
+
+  TORCH_INTERNAL_ASSERT(
+      current_operand_.has_value() && "Unexpected operator type");
+}
```
This is dead code.

prak-nv commented on torch/csrc/jit/codegen/cuda/kernel_expr_evaluator.cpp:
```diff
@@ -3,14 +3,166 @@
+  hashmap_t* known_values_ = nullptr;
+};
+
+void ExprBuilderVisitor::visit(const BinaryOp* binary_op) {
```
Again, currently unused

prak-nv replied:
Removed

prak-nv committed:
WIP: Use gldfc expression evaluator for runtime shape inference

To speed up expression shape expressions for memory allocation and
parallel extenst we memoize such expressions in runtime inside of
KernelSummary. Expressions are evaluated during runtime when all runtime
values are already known.

naoyam commented:
I haven't reviewed the glfdc directory very closely yet. I think I first need your help to understand some of the basic concepts. Please see my question below

> I'm getting confused with the expression-related classes in `namespace glfdc`. Can you explain what they represent and how they should be used?
> 
> * glfdc::Expr
> * glfdc::ExprDag
> * glfdc::SExpr
> * glfdc::SExprRef

prak-nv commented:
> I'm going to review this PR. For the meantime, please consider adding tests. It might make sense to have a new test file, like "test_gpu_glfdc.cpp", in the test directory.

Still working on porting my tests to gtest, will address that in follow up change

prak-nv commented:
> I haven't reviewed the glfdc directory very closely yet. I think I first need your help to understand some of the basic concepts. Please see my question below
> 
> > I'm getting confused with the expression-related classes in `namespace glfdc`. Can you explain what they represent and how they should be used?
> > 
> > * glfdc::Expr
> > * glfdc::ExprDag
> > * glfdc::SExpr
> > * glfdc::SExprRef

* glfdc:ExprDAG - container for all expressions - those are kept inside `vector<SExpr>`, subexpression describe single binary operations with integer scalars (symbolic or known) or other nested subexpressions. The subexpressions can be only added at the end of this vector and are never reordered. This makes possible to refer to given subexpression by its index into this vector. Responsibilty of adding subexpressions to DAG is delegated to `glfdc::ExpressionBuilder`
* glfdc::SExpr describes single subexpression (binary operation) and references to its operand. It has strict order for comparison and is comparable by value. glfdc::SExpr are also immutable during whole lifetime of owner `glfdc::ExprDAG`
* glfdc::SExprRef is an index into array of subexpression and is valid as long owner `glfdc::ExprDAG` exist.
* glfdc::Expr - is convinence class that holds reference to `glfdc::SExpr` and its owner `glfdc::ExprDAG`.

Please let me know if comments in glfdc for those types should be improved to explain those concepts better.

prak-nv commented on torch/csrc/jit/codegen/cuda/kernel_expr_evaluator.cpp:
```diff
@@ -23,23 +58,33 @@ void ExpressionEvaluator::bind(
     return value->as<Int>()->value();
   }
 
-  // Is the value known (either explicit binding or memoized)?
```
This comment should be reverted back

prak-nv commented:
Dropped #998 change as it is not required or directly related.

naoyam commented on torch/csrc/jit/codegen/cuda/shape_expr_memo.cpp:
```diff
@@ -0,0 +1,123 @@
+using fuser_cuda::glfdc::ExprDAG;
+using fuser_cuda::glfdc::ExprEvaluator;
+using fuser_cuda::glfdc::Operand;
+using fuser_cuda::glfdc::Value;
```
These `using` seem to make the code bug prone. For example, `Val` is `kir::Val`, but `Value` is `glfdc::Value`, and that seems too subtle to notice and remember. And since this is in the fuser namespace, `Val` may actually mean the fuser `Val`. For safety and clarity, I'd suggest the `kir` and `glfdc` prefix should not be dropped except it's extremely clear.

naoyam commented on torch/csrc/jit/codegen/cuda/lower2device.h:
```diff
@@ -14,6 +14,18 @@
 #include <memory>
 #include <ostream>
 
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+namespace kir {
+class ExprBuilderVisitor;
+}
+} // namespace cuda
+} // namespace fuser
+} // namespace jit
+} // namespace torch
+
```
nit: We could just have `namespace kir { class ExprBuilderVisitor; }` inside the cuda namespace available below. That would make it a little clearer that it shares the same top-level namespace.

naoyam commented on torch/csrc/jit/codegen/cuda/kernel_expr_evaluator.h:
```diff
@@ -53,6 +66,7 @@ class TORCH_CUDA_CU_API ExpressionEvaluator : private IrVisitor {
 
  private:
   std::unordered_map<const Val*, Int::ScalarType> known_values_;
+  std::unique_ptr<glfdc::EvalState> eval_state_;
```
Please add a comment here.

naoyam commented on torch/csrc/jit/codegen/cuda/glfdc/sexpr.h:
```diff
@@ -0,0 +1,148 @@
+// Operand of subexpression
+using Operand = c10::variant<c10::monostate, Value, SExprRef>;
+
+// Returns true IFF operand is know or unknown (symbolic) scalar
```
typo: known

naoyam commented on torch/csrc/jit/codegen/cuda/glfdc/bitvector.cpp:
```diff
@@ -0,0 +1 @@
+#include <torch/csrc/jit/codegen/cuda/glfdc/bitvector.h>
```
Do we need this file?

prak-nv replied:
Not really, beyond checking header compiles standalone.

shmsong replied:
I'd also prefer not having these cpps to save some object files in the build : ).

naoyam commented on torch/csrc/jit/codegen/cuda/glfdc/expr.h:
```diff
@@ -0,0 +1,80 @@
+};
+
+// Keeps DAG of all expressions reduction graphs
+struct ExprDAG {
```
nit: structs are used only for trivial cases in our codebase, so `class` is recommended here.

prak-nv replied:
Will do, usually I just picked one that gives me less public/private clauses.

naoyam commented on torch/csrc/jit/codegen/cuda/glfdc/expr.cpp:
```diff
@@ -0,0 +1,24 @@
+  TORCH_INTERNAL_ASSERT(ref.index() < reused_subexpressions_.size());
+
+  bool was_reused = reused_subexpressions_[ref.index()];
+  reuse_count_ += std::size_t(was_reused ^ true);
```
Isn't this easier to understand?

```
if (!was_reused) {
  ++reuse_count_;
}
```

piotrrak replied:
Possibly :)

naoyam commented:
> > I haven't reviewed the glfdc directory very closely yet. I think I first need your help to understand some of the basic concepts. Please see my question below
> > > I'm getting confused with the expression-related classes in `namespace glfdc`. Can you explain what they represent and how they should be used?
> > > 
> > > * glfdc::Expr
> > > * glfdc::ExprDag
> > > * glfdc::SExpr
> > > * glfdc::SExprRef
> 
> * glfdc:ExprDAG - container for all expressions - those are kept inside `vector<SExpr>`, subexpression describe single binary operations with integer scalars (symbolic or known) or other nested subexpressions. The subexpressions can be only added at the end of this vector and are never reordered. This makes possible to refer to given subexpression by its index into this vector. Responsibilty of adding subexpressions to DAG is delegated to `glfdc::ExpressionBuilder`
> * glfdc::SExpr describes single subexpression (binary operation) and references to its operand. It has strict order for comparison and is comparable by value. glfdc::SExpr are also immutable during whole lifetime of owner `glfdc::ExprDAG`
> * glfdc::SExprRef is an index into array of subexpression and is valid as long owner `glfdc::ExprDAG` exist.
> * glfdc::Expr - is convinence class that holds reference to `glfdc::SExpr` and its owner `glfdc::ExprDAG`.
> 
> Please let me know if comments in glfdc for those types should be improved to explain those concepts better.

Thanks. That's very helpful, now I see what the overall system looks like more clearly. 

Please make sure to have these notes available somewhere in the code or perhaps in a separate design doc.

naoyam commented on torch/csrc/jit/codegen/cuda/glfdc/expr.h:
```diff
@@ -0,0 +1,80 @@
+    return reuse_count_;
+  }
+
+  // Marks subexpression as reused for CSE
```
What does it mean when a subexpression is reused?

prak-nv replied:
It was marked as common part of another expression by CSE. As a simple example, given expressions `(x+1)*7` and `(x+1)*y` the `(x+1)` subexpression is common part. Once evaluated once for given value of x the result is kept in glfdc::EvalState slot and doesn't need to be ever evaluated.

naoyam replied:
Please leave it as a comment.

naoyam commented on torch/csrc/jit/codegen/cuda/glfdc/expr.cpp:
```diff
@@ -0,0 +1,24 @@
+  reused_subexpressions_[ref.index()] = true;
+  auto e = this->fetch(ref);
+
+  if (!isValue(e.lhs_))
```
Wouldn't it be safer to use `isSExpr` instead of `isValue` as `!isValue` does not mean `isSExpr`?

naoyam commented on torch/csrc/jit/codegen/cuda/kernel_expr_evaluator.cpp:
```diff
@@ -3,14 +3,49 @@
+  using hashmap_t = std::unordered_map<const Val*, Int::ScalarType>;
+
+ public:
+  c10::optional<Int::ScalarType> operator()(uintptr_t binding) {
```
Why is `uintptr_t` used? Why not just `const Val*`?

prak-nv replied:
At the cost of generality of using glfdc for different types, we could simply use kir::Val*.
The only assumption we have right now is that it is unique value for different symbolic values.

naoyam replied:
Thanks. Please leave it as a comment somewhere.

naoyam commented on torch/csrc/jit/codegen/cuda/glfdc/eval.h:
```diff
@@ -0,0 +1,182 @@
+using opt_index_t = c10::optional<std::size_t>;
+
+// Holds mapping of reused subexpressions
+struct ReusedExprMapping {
```
Use class rather than struct.

naoyam commented on torch/csrc/jit/codegen/cuda/glfdc/eval.h:
```diff
@@ -0,0 +1,182 @@
+    for (std::size_t i = 0; i < reused_sexprs.size(); ++i) {
+      // if subexpression is reused assign mapping
+      if (reused_sexprs[i])
+        mapping.insert(i, slot_counter++);
```
It is generally preferred to have braces always, so:

```
if (reused_sexprs[i]) {
   mapping.insert(i, slot_counter++);
}
```

piotrrak replied:
Sure. Then, should we add .clang-format with such rule in codegen/cuda directory?

naoyam replied:
Yeah, that would be even better.

prak-nv replied:
It seems, clang-format can't do that, clang-tidy can. Guess it'll be up to me running it pre-submit. For reference `git diff -U0 HEAD^ | clang-tidy-diff.py -fix  -checks="-*,readability-braces-around-statements" -p1`

naoyam commented on torch/csrc/jit/codegen/cuda/glfdc/eval.h:
```diff
@@ -0,0 +1,182 @@
+  // value slot index
+  static ReusedExprMapping createLazyMapping(
+      const BitVector& reused_sexprs) // O(n)
+  {
```
The convention here places `{` directly after the parameter declaration without a newline. I think clang-format gets confused with the comment (`// O(n)`) in this particular case.

naoyam commented on torch/csrc/jit/codegen/cuda/glfdc/eval.h:
```diff
@@ -0,0 +1,182 @@
+// prepared and memoized) During evaluation runtime (unknown) values are being
+// queried and inserted to the stack copy Afterwards expression is calculated by
+// DFS right first tree reduction using stack algorithm
+struct ExprEvaluator {
```
Please use class.

naoyam commented on torch/csrc/jit/codegen/cuda/glfdc/eval.cpp:
```diff
@@ -0,0 +1,336 @@
+
+namespace {
+
+using skiptree_t = std::vector<std::pair<SExprRef, std::size_t>>;
```
Please add a comment here. What does each pair represent?

shmsong replied:
I think this type is basically a tree structure flattened out on a vector. In each vector element `skip_tree[i]`, `skip_tree[i].first` is a reference to the actual tree/IR node, and `skip_tree[i].second` is the incremental index on the vector. For example, all child nodes of `skip_tree[i].first` indexed on `skip_tree` will be the nodes between `i` and `i+skip_tree[i].second`, i.e `skip_tree[i+1]` to `skip_tree[i+skip_tree[i].second]`

@prak-nv please correct if not accurate and please do include comments explaining this data structure.

piotrrak replied:
Yes, skip_tree has same structure as list of operations. In other words is skip_tree[i].second is number of childs in subtree where skip_tree[i].first is root node.

naoyam commented on torch/csrc/jit/codegen/cuda/glfdc/eval.cpp:
```diff
@@ -0,0 +1,336 @@
+
+  list[index].second = lchild_count + rchild_count;
+
+  return 1 + lchild_count + rchild_count;
```
This seems like an important function. Please leave detailed comments.

naoyam commented on torch/csrc/jit/codegen/cuda/glfdc/eval.cpp:
```diff
@@ -0,0 +1,336 @@
+    skiptree_t& list,
+    FnPre_& visit_pre,
+    FnPost_& visit_post,
+    LRGetters_& getters) {
```
Do these parameter types really need to be templatized? Is it really necessary to templatize the logic to access tree nodes? It seems to me that the interface is overly generalized, making a simple function non-straightforward to follow. Unless it really needs to be, I'd prefer to make it more trivial and minimally generic.

piotrrak replied:
Don't need to be.

naoyam commented on torch/csrc/jit/codegen/cuda/glfdc/expr_builder.h:
```diff
@@ -0,0 +1,69 @@
+  // Returns symbolic value for unknown that can be used in subexpression
+  Value getBinding(uintptr_t unbound); // O(log2(n))
+
+  Operand createSExpr(OperatorKind op, Operand l, Operand r) {
```
Please add a comment.

naoyam commented:
I'm done reading all your code. It's really nice work. Although I'm not yet 100% confident about technical details, I'm pretty sure the runtime evaluation cost should be greatly reduced.

Here're my major suggestions:

- More detailed and thorough comments. There's nothing wrong with having verbose comments. It may be obvious once you get the whole picture, but that could be really helpful for someone reading the code first time.
- To make it easier to understand the code, please try to simplify the code when appropriate. For example, does `makeDFSTraversal` need to create a function? Why can't it just do traversal?
- Tests. Since glfdc introduces a completely separate arithmetic system, there're a lot of cases that must be tested.

Also, I'm curious what the performance looks like. Do you have performance results?

shmsong commented on torch/csrc/jit/codegen/cuda/executor.cpp:
```diff
@@ -329,44 +353,62 @@ uint64_t FusionExecutor::computeSharedMemory(
+uint64_t FusionExecutor::computeSharedMemory(
+    kir::ExpressionEvaluator& expr_eval,
+    const std::vector<const kir::Allocate*>& buffers,
+    const std::vector<glfdc::ExprEvaluator>& memoized,
```
`Memoized` looks like logic on the `Int`'s instead of the actual allocation or shared mem computation. Does it make sense to wrap the memoized argument in `kir::ExpressionEvaluator`? Or maybe some other alternatives to having to provide a memoized argument to all functions that could benefit from memoization.

piotrrak replied:
I will have to think bit more about it. I've already did give it some thought, since it would be nice to somehow integrate it better with kir altoghter.

shmsong commented on torch/csrc/jit/codegen/cuda/glfdc/cfold.cpp:
```diff
@@ -0,0 +1 @@
+#include <torch/csrc/jit/codegen/cuda/glfdc/cfold.h>
```
Same here. Would prefer not to have header-only cpp's

shmsong commented on torch/csrc/jit/codegen/cuda/glfdc/cfold.h:
```diff
@@ -0,0 +1,71 @@
+    return c10::get<scalar_type>(c10::get<Value>(op));
+  };
+
+  if (!isValue(v1) || isUnboundValue(v1))
```
Just curious, how simple is it to check what values were `unbound` in this interface? It could be helpful if some error message shows what values were unbound when a requested expr doesn't evaluate.

piotrrak replied:
`isUnboundValue()` simply checks which member of discriminated union is active ie `std::variant<>::index()` member.
It should be O(1) operation assuming you've got access to owner DAG. Checking if ExprEvaluator has any unbound values, would be O(n) where n is numer of terms-1 in expression.
 `dag.unbound_values[std::get<UnboundValue>(v1).index_]` gives you value that binding was created.
As for ExprEvaluator the `binding_gaps_` vector is esentially list of all unbound values of an expression.

shmsong replied:
I see. Thanks for the clarification. 👍

shmsong commented on torch/csrc/jit/codegen/cuda/glfdc/eval.cpp:
```diff
@@ -0,0 +1,336 @@
+    FnPre_& visit_pre,
+    FnPost_& visit_post,
+    LRGetters_& getters) {
+  list.emplace_back(t, 0);
```
Just some addition to @naoyam's comment on implementation:

While functional implementations are interesting, we tend to lean more towards OOP. The `fn`/`callback`'s passed in this generic traversal are actually callable objects that carry some states. It may also make sense to wrap them in classes. I think it'd be great if callable parameters could be replaced with objects with concrete names, like `ChildExprGetter`, and actual methods could be used instead of `operator ()`, like `child_expr_getter.getLeftChild()`.

piotrrak replied:
Will do.

shmsong commented on torch/csrc/jit/codegen/cuda/glfdc/eval.cpp:
```diff
@@ -0,0 +1,336 @@
+  auto left = left_fn(t);
+  size_t lchild_count = 0, rchild_count = 0;
+
+  visit_pre(t);
```
It'd be great if we could make this function less generic and we could discuss more about a generic implementation if we have multiple use cases of the same pattern. 

If I'm understanding correctly, this `dfs` function is used to initialize the stack. `Visit_pre` in its instantiation places the left child of `t` on stack while `visit_post` places the right child of `t` on stack. I understand the ordering consideration. If this is correct I think this function would be easier to understand if `visit_pre` and `visit_post` could be replaced with the concrete operation it does in the context, if that's not too much update.

shmsong commented on torch/csrc/jit/codegen/cuda/glfdc/eval.cpp:
```diff
@@ -0,0 +1,336 @@
+    typename FnPre_,
+    typename FnPost_,
+    typename LRGetters_>
+auto makeDFSTraversal(
```
Another functional vs OOP, I'd really appreciate it if this could be replaced with a factory and return type could use a class that has a traversal method instead of lambda.

piotrrak replied:
Sure

shmsong commented on torch/csrc/jit/codegen/cuda/glfdc/eval.cpp:
```diff
@@ -0,0 +1,336 @@
+    // Coerce right subtree root index
+    rroot_idx = lroot_idx + 1 + l_subop.nsubops;
+
+    operand_count += calculateOperandsRange(lroot_idx, rroot_idx);
```
@naoyam , it looks like in other parts of the fuser we prefer pseudo recursion and I'm slightly worried about the stack these trees will produce. Would you have suggestions on this? Also, @prak-nv would it be difficult/harmful to use pseudo recursion in some of these implementations?

piotrrak replied:
Had TODO for reworking whole thing to iterative DFS version. Recursive is always more straightforward in my eyes and given the expressions we have in practice aren't too deep, I've decided not to. It should be doable. This code affects only compile time and hit on stack allocations for DFS should be negligable. I wouldn't go for psudo-recursion though.

naoyam replied:
I don't have any preference. I agree with Piotr that it probably won't matter much as expressions don't become very deep.

shmsong replied:
I see. I think keeping the recursive version should be fine if we can assume the trees won't be very deep.

shmsong commented on torch/csrc/jit/codegen/cuda/glfdc/eval.cpp:
```diff
@@ -0,0 +1,336 @@
+  // This is still DFS but of mirrored tree (or right child first)
+  //
+  // Right needs evaluation?
+  if (!isValue(root_expr.rhs_))
```
nitpick: curly bracket.

shmsong commented on torch/csrc/jit/codegen/cuda/glfdc/eval.cpp:
```diff
@@ -0,0 +1,336 @@
+
+ExprEvaluator::opt_scalar_t ExprEvaluator::evaluate(
+    EvalState& es,
+    binding_fn_t binding_fn) const {
```
Same here, the `binding_fn` carries state so it'd be great if a class instead of a function is passed here and methods with names were used instead of `operator ()`

piotrrak replied:
Would templatizing it in addition to that be also OK? I am already bit worried `std::function::operator()` is virtual and that std::function constructor must do allocation. For now I've let it slip, until I get profile on it.

shmsong replied:
Sure I agree that performance profile should largely drive the decision here and I'm not against templatizing if it's perf critical. 

Also just curious is it necessary to pass a `std::function` as the "binding funciton"? It looks like its trying to fill in values for the unbound variables so could we just pass a reference to a book-keeping class like `KnownValues`? It'd just be easier to follow.

piotrrak replied:
This is really just an artifact of it being developed as standalone thing to prove the concept. https://github.com/prak-nv/glfdc

shmsong commented on torch/csrc/jit/codegen/cuda/glfdc/expr.h:
```diff
@@ -0,0 +1,80 @@
+  }
+
+  // Returns number of reused subexpression inside DAG
+  std::size_t reuseCount() const noexcept {
```
I might have missed it but is the `reuse_count_` used?

piotrrak replied:
For now, it is not. But it is important metric for me and useful in unit tests (still WIP).

shmsong commented on torch/csrc/jit/codegen/cuda/glfdc/sparse_map.h:
```diff
@@ -0,0 +1,87 @@
+// described https://research.swtch.com/sparse
+//
+// Uses same trick but stores both k,v pair in dense array
+class sparse_map {
```
This implementation is great. Just a curious question: can it support `erase` as well?

piotrrak replied:
Sure, we can support it, also O(1) operation.

shmsong commented on torch/csrc/jit/codegen/cuda/glfdc/stack.h:
```diff
@@ -0,0 +1,85 @@
+  // drop - drops n operands from stack
+  void drop(std::size_t n) noexcept {
+    TORCH_CHECK(n <= this->c.size());
+    this->c.resize(this->c.size() - n);
```
I'm not sure how confident we are with the stability of this protected variable of `std::stack` over time. Is there any doc from `libstdc++` or `libc++` about this `c` existing long term? The implementation makes sense but just the protected variable naming `c` from `std::stack` did not sound particularly user-facing. cc @naoyam @csarofeen 

Also since `std::stack` is supposed to be a wrapper class, would it make sense to include our "local stable" version in the repo instead of inheriting the `std` one?

piotrrak replied:
Since this protected member it's been part of HP STL released in 1994 and been part of standard to this day (c++20) I doubt we should be concerned about it. Please see N4868 `22.6.6.2 [stack.defn]`. True, it might be deprecated in c++23 and then removed in c++2z but that can be said about anything else.

shmsong reviewed (commented):
This is very interesting and promising work. Had some comments on code structure.

piotrrak commented:
> * More detailed and thorough comments. There's nothing wrong with having verbose comments. It may be obvious once you get the whole picture, but that could be really helpful for someone reading the code first time.
Will work on that.
> * To make it easier to understand the code, please try to simplify the code when appropriate. For example, does `makeDFSTraversal` need to create a function? Why can't it just do traversal?
It can, sorry about templetes lambdas, intent was not meant to be make those generic, but to make it clear what algorithms are used for complexity analysis. 
> * Tests. Since glfdc introduces a completely separate arithmetic system, there're a lot of cases that must be tested.
Right now working on porting my Catch2 testsuite to gtest
> 
> Also, I'm curious what the performance looks like. Do you have performance results?
Thats also WIP unfortunely, curious about it too.

prak-nv commented on torch/csrc/jit/codegen/cuda/glfdc/sexpr_cmp.h:
```diff
@@ -0,0 +1,91 @@
+
+// Boost hash_combine();
+template <typename Ty_>
+inline void hash_combine(std::size_t& seed, const Ty_& v) noexcept {
```
c10 has hash_combine on its own, this can be replaced by their version

prak-nv commented on torch/csrc/jit/codegen/cuda/glfdc/cfold.h:
```diff
@@ -0,0 +1,71 @@
+      return l * r;
+    case OperatorKind::div: {
+      if (r == 0)
+        return 0;
```
Not to sure it should 0 division should be defined, this should rather assert.

prak-nv commented on torch/csrc/jit/codegen/cuda/glfdc/cfold.h:
```diff
@@ -0,0 +1,71 @@
+    }
+    case OperatorKind::ceildiv: {
+      if (r == 0)
+        return 0;
```
Also here.

prak-nv commented on torch/csrc/jit/codegen/cuda/glfdc/cfold.h:
```diff
@@ -0,0 +1,71 @@
+    }
+    case OperatorKind::mod: {
+      if (r == 0)
+        return 0;
```
And here

shmsong commented on torch/csrc/jit/codegen/cuda/executor.cpp:
```diff
@@ -235,9 +235,13 @@ void FusionExecutor::compileFusion(
 at::Tensor inferAndAlloc(
     const kir::TensorView* tv,
     const std::vector<kir::Val*>& sizes,
+    OptMemoizedExpressions opt_memoized_sizes,
```
It'd be great if `inferAndAlloc`, `computeLaunchParams` and `computeSharedMemory` would be able to use the optimized expr_eval without knowing about the optimizations/memoizations. These functions have some complexity and may drift over time so it'd be nice to keep them orthogonal to the integer evaluators. 

I'd prefer maintaining an interface like in #1086 that all these functions had to do was just to call `expr.evaluate(...)`.

shmsong commented on torch/csrc/jit/codegen/cuda/executor.cpp:
```diff
@@ -246,15 +250,41 @@ at::Tensor inferAndAlloc(
-        " for the buffer ",
-        kir::toString(tv));
-    inferred_sizes.push_back(inferred_val.value());
+  if (opt_memoized_sizes.has_value()) {
```
Similarly, I guess this part of logic could go inside of the evaluator itself. We wouldn't want `inferAndAlloc` to take on too much work on optimized integer evaluation.

shmsong commented on torch/csrc/jit/codegen/cuda/executor.cpp:
```diff
@@ -328,61 +358,90 @@ uint64_t FusionExecutor::computeSharedMemory(
+  const auto& kernel_summary = kernel()->summary();
 
   auto data_cache = compileTimeDataCache();
-
```
I'm relatively open on this implementation and we can discuss more.

Putting these in either `KernelSummary` or `ExecutorCompileTimeEntry` wouldn't make much difference timing-wise: 

We discussed earlier that `ExecutorCompileTimeEntry` would be computed first time used while `KernelSummary` is computed in `GpuLower`. But generally we lower/compile these fusion graphs the first time we run it so the latency of the first graph launch will be combined with lowering and nvrtc compile either way. 

For the discussion, my thinking behind creating the `ExecutorCompileTimeEntry` while we already had `KernelSummary` was that I thought we'd need a caching space for pure performance improvement and the `data we put in the ExecutorCompileTimeCache` are:

1. Not particularly as "meaningful" as what we have in `KernelSummary`. They could be just some vector of pointers that we wouldn't pay particular attention to when lowering the graph. We just need them computed and cached so that we don't need to do it every time we run the kernels. 

2. and because of 1., some of them are subject to change in the future, possibly quite frequently, so the entry definitions were more ad hoc/flexible and I thought it'd be helpful to place the computation of the entries exactly in-place where they're needed.

But either way, we could always do follow up PR's if we think one way is particularly cleaner looking than the other.

shmsong commented on torch/csrc/jit/codegen/cuda/glfdc/eval.cpp:
```diff
@@ -0,0 +1,220 @@
+SymbolicExpr::SymbolicExpr(const Expr& e)
+    : expr_(e), templ_(SymbolicExprVisitor::createTemplate(e, stack_order_)) {}
+
+void SymbolicExpr::evaluateSubExpr(
```
Do we need this evaluation method if we already have `UProgram`?

shmsong commented on torch/csrc/jit/codegen/cuda/glfdc/eval.h:
```diff
@@ -0,0 +1,283 @@
+
+// OperandStackOrder - determines order of operand on stack, ie. evaluation
+// order left/right side first
+enum class OperandStackOrder {
```
Why do we need to implement 2 stack orders?

shmsong commented on torch/csrc/jit/codegen/cuda/glfdc/eval.h:
```diff
@@ -0,0 +1,283 @@
+
+  // Evaluates expression in optimized form
+  opt_scalar_t evaluateOptimized(EvalState& es, binding_fn_t binding_fn) const;
+  // Evaluates expression from template (fallback)
```
When is template used? Looks like if the optimized form doesn't have the value we want. Would we be able to go back to legacy pointer chasing method on the IR nodes if that's the case?

shmsong commented on torch/csrc/jit/codegen/cuda/glfdc/eval_vm.h:
```diff
@@ -0,0 +1,734 @@
+  scalar_type a, b, t;
+
+  // Stack of operands
+  OperandsStack operands(memory);
```
Just curious if there's any particular reason for having 3 registers and 2 stacks. Is it because of more convenient register allocation or better support of control flows?

shmsong commented on torch/csrc/jit/codegen/cuda/glfdc/expr.cpp:
```diff
@@ -0,0 +1,65 @@
+
+#include <c10/util/Exception.h>
+
+using namespace torch::jit::fuser::cuda::glfdc;
```
nitpick: we'd want to put the implementations in namespace `torch::jit::fuser::cuda::glfdc` instead of pulling the symbols into `::`

shmsong commented on torch/csrc/jit/codegen/cuda/glfdc/eval_vm.h:
```diff
@@ -0,0 +1,734 @@
+  b = 0;
+  t = 0;
+
+  for (; ip < instructions.size(); ++ip) {
```
It seems like having these 3 registers, a scratch space, and 2 orders of stack visit is creating quite a few copies of the same instructions. Just wondering if we could simplify this instruction set a bit.

shmsong reviewed (commented):
Had a few inline questions. Also questions on the overall design:

1. Could you give a summary of the usage of expression evaluation? I see at least 3 representations now, we have `kir::Expr`, `glfdc::ExprDAG`, and `UProgram`. It'd be nice to have an overview of how these layers can be used and how they interact.  

2. Could you give a bit overview on the design decisions of the eval_vm? Particularly interested in the design of register and stack space and the choice of having different versions of the same instruction for different src/dst.
